### ISSUE TITLE
Installer problems when building against v1.8.2-vmp1010

### ISSUE THREAD

**Issue by simongregory** [castlabs/electron-releases#8]
Our app uses the built in Electron updater which means we run it through the [windows-installer] tool to build the binaries we ship. When we use the win32 x64 build of v1.8.2-vmp1010 the installer and setup binaries are built without error, however on running the setup.exe it warns of a missing `VCRuntime140.dll` file:

![install-problem-small](https://user-images.githubusercontent.com/33640/37032091-305eb108-2139-11e8-980a-075dd5986e81.png)

This doesn't happen with the vanilla build of Electron v1.8.2, nor with the 32bit version of v1.8.2-vmp1010. There's not much in the way of help in the Squirrel logs so I'm at a loss as to why we get the error - as the file looks to get bundled.

One thing I've noticed is that if I md5 the x64 `.dll` files in the Castlabs build and compare against the vanilla Electron build there are 2 files that are the same - `vcruntime140.dll` being one of them.

Any thoughts would be appreciated, I'll comment if I find anymore info or a resolution.

[windows-installer]: https://github.com/electron/windows-installer

**Comment by simongregory** [comment]
Thanks for the fast response Emil. I'd expected that the dll's were shared but thought it was worth mentioning.

The build machine I'm using is a Windows 10 VM, with the essentials installed to generate our releases. When I checked there was no `Microsoft Visual C++ 2015 Redistributable` but there were `2012` entries of the same name for both x64 and x86. These have been recently installed/updated (presumably as part of a system update). If there was a problem with a library dependency for the installer I would expect to see the same behaviour for both the VMP and vanilla builds of Electron - however I will explore this and install the 2015 version.

My (limited) understanding of the installer exe is that it is only dependent on the .NET framework and when it is not found will download it before proceeding.

Although I've not found a way to unpack the setup.exe to be sure, I have examined the NuGet `.nupkg` file (it's just a zip) that's generated at the same time and used in the auto-update mechanism and it does have the `vcruntime140.dll` bundled inside.

**Comment by khwaaj** [comment]
Thanks for the additional information. I've not been able to reproduce any of this as of yet, in fact I'm having trouble generating a working installer at all using `electron-winstaller` (we have mostly been working with `electron-builder` for our own tests).

Yeah, I temporarily forgot that the vanilla Electron worked as expected, presumably generated on the same machine. This makes my suggestion above look very unlikely, although this all appears a bit odd to begin with, since there should really be no difference here, so it is difficult to say anything for sure right now.

BTW, The Downloads.exe binary (in the dialog title bare above), is that the actual Electron application, or something else?

**Comment by simongregory** [comment]
So you're aware we use a private fork of [windows-installer](https://github.com/electron/windows-installer) which is modified to include vmp signing, bundle the .sig files, and customise the app icon. Before raising this I did ensure that the same behaviour was exhibited using the public version without the modifications.

The 'Downloads.exe' is our Electron app which is getting launched as part of the installation (I should have been clearer about that). On disk it is missing the `vcruntime140.dll` in `AppData\local\downloads\app-2.5.4\`. So far I've not been able to determine any other differences.

I've been suspicious from the start that the symptom (the missing .dll) is a side effect of an error elsewhere. We have tried the installer on alternative machines (including real hardware), and get the same results. Thinking about it I do need to ensure that those machines haven't got Windows Studio installed incase the installer is designed to skip the .dll if it finds it on the machine (or mistakenly thinks it can find it).

We have had problems reported with the .dll's previously, 32 bit users have reported that deleting the `d3dcompiler_47.dll` file has stopped 'white screen' issues. Which suggests that the installer is designed just to install all the files it finds bundled.

**Comment by khwaaj** [comment]
I managed to reproduce your issue, after a fashion.

It appears that the generated nupkg archive contains everything it should, but the extracted contents in AppData does not, which leads me to believe the error occurs during the unarchiving process. The `resources/widevinecdmadapter-win-x64.zip` file is the first one to exhibit problems in my case (it has a zero size after unarchiving) and any file that would alphabetically come after that file in the archive is not present in the final unarchived structure, which includes `vcruntime140.dll` since it will come after any file in `resources/`. Unarchiving the package using regular zip tools seems to work fine, so this appears to be specific to the implementation used by the installer.

I've not yet found a solution but I'll keep digging tomorrow. Maybe you have some ideas based on my findings, you are probably more familiar with the installer implementation than I am since you have forked and worked on it a bit.

**Comment by khwaaj** [comment]
`resources/widevinecdmadapter-win-x64.zip` is definitely the culprit here. If I extract the contents and then recreate it using 7z on Windows and use that in place of the original zip the installation works as it should. I'm not sure what to do about it though since the original zip seems to work fine in all other cases I've encountered, so this appears to be some incompatibility between the zip implementations used to package our Electron and the one in the NuGet installer.

Would you try replacing your zip-file with the one I'm attaching to make sure it is the same issue we are seeing? You can also remove `widevinecdmadapter-win-ia32.zip` from your package to make sure that is not causing trouble (it is not necessary in the x64 build anyway and will go away in future releases).

[widevinecdmadapter-win-x64.zip](https://github.com/castlabs/electron-releases/files/1788744/widevinecdmadapter-win-x64.zip)

**Comment by simongregory** [comment]
Thanks Emil - and good spot on the zip! I can confirm I'm seeing the same behaviour here. The first thing that occurred to me was to remove the cdmadapter ia32 zip from the 64bit build and try again.

Am just working through setting the download mirror up to serve the adjusted files to the build process - but zipping the release up on a mac appears to have upset the build, this shouldn't be too much trouble to sort out. I'll swap in your re-zipped adapter as well and let you know.

**Comment by simongregory** [comment]
Swapping in the zip above works, thanks!

I'm not exactly sure which upstream project is responsible for the zip compatibility issue, but [Squirrel.Windows](https://github.com/Squirrel/Squirrel.Windows) should not have silently failed. I'll raise an issue there so the problem is at least discoverable incase anyone else encounters it.

**Comment by khwaaj** [comment]
Nice, that is good news. I'll try to make a new Electron release this week since 1.8.3 has been released, including a security fix, so I'll take a look at the step zipping this up to see if there is something we can do on our side to work around this. It may be a bit difficult though since it only appears to affect some unknown subset of the zips we produce, maybe it'll just work out of the box for this build.

**Comment by khwaaj** [comment]
Actually, after thinking a bit more about this, I don't think there is much to do on our side of things. The file is what it is and should not affect it being stored inside another archive, regardless of contents.

One thing of note though that you may want to include in your upstream report: The original CDM zip is encoded as `stored` (uncompressed) within the nuget package, whereas the re-zipped one gets encoded as `deflated` (compressed), so that is likely to be of significance. There are other files/directories that are also `stored` within the archive, but none that is very large, so maybe it is a size-related issue for `stored` items.

**Comment by simongregory** [comment]
I agree, the zip really should just be treated as arbitrary data and copied into place by the installer.

Thanks for your help again Emil, the re-zip of the cdmadapter gets us unblocked. I'll keep an eye out for 1.8.3 and hope the zip just works in that build.

**Comment by khwaaj** [comment]
I just published our 1.8.3 release and I tested it with the installer setup I used when investigating this. The good news is that it works out of the box, now we'll just have to hope this holds true for future releases as well (or even better, that the issue gets fixed upstream).

BTW, you mentioned an issue with ´d3dcompiler_47.dll´. I looked into this briefly and found that some of our 32-bit releases actually contains a 64-bit version of this DLL, which could very well be the cause of this issue. I've made some updates to the build-system to avoid this happening in the future.

**Comment by simongregory** [comment]
Thanks Emil! I'd spotted it and am part way through updating the build, I spotted the zip had changed so had my fingers crossed. Glad it's going to be working out of the box, and hopefully it'll continue to.

On the `d3dcompiler_47.dll` issue we'd not been able to get a repo in house, and there were issues with the same symptoms in the Electron repo so I'd concluded it was something upstream. Will mention it if we get anything similar in the future. Thanks for mentioning it.

**Comment by devaffair** [comment]
We're experiencing (what I believe is) the same issue and was wondering if you got a solution.  Our electron dependency is:
`"electron": "https://github.com/castlabs/electron-releases#v1.8.3-vmp1010",`

and our command is:
`electron-packager . --overwrite --platform=darwin --arch=x64 --icon=src/assets/icons/mac/icon.icns --prune=true --out=release-builds`

The error we receive is:
`Downloading tmp-14712-0-electron-v1.8.3-vmp1010-darwin-x64.zipError: GET https://github.com/electron/electron/releases/download/v1.8.3-vmp1010/electron-v1.8.3-vmp1010-darwin-x64.zip returned 404
Failed to find Electron v1.8.3-vmp1010 for darwin-x64 at https://github.com/electron/electron/releases/download/v1.8.3-vmp1010/electron-v1.8.3-vmp1010-darwin-x64.zip`

Any thoughts about how can this be sorted out?
Thanks.

**Comment by simongregory** [comment]
@devaffair, this isn't the same issue but try configuring `electron-packager` to use a download mirror - you can see in the error it's trying to download (via `electron-download`) from the main electron repo.

These `options` work for us when invoking the packager for Windows (for legacy reasons our Mac build is different):

```
  download: {
    mirror: 'https://github.com/castlabs/electron-releases/releases/download/v',
    cache: './.electron-dl-cache'
  },
```

**Comment by obecny** [comment]
@devaffair @simongregory the `electron-packager` is by default using the electron version from `electron` not from Castlabs. That's why you need to overwrite this settings using `package.json` for example adding this option
```
  "build": {
    "electronDist": "node_modules/electron/dist/"
  },
```
Then instead of downloading this again it will simply use the version you already declared in your `package.json` - the Castlabs version.

**Comment by devaffair** [comment]
@obecny  So, if I understand correctly, I need to modify package.json and add
` "build": {
    "electronDist": "node_modules/electron/dist/"
  },`

Should that be added in the application's package.json? or in the electron-packager's package.json?

I've tried on both with the same error, and it seems like there's something I'm missing here..

**Comment by obecny** [comment]
@devaffair `electronDist` is a part of electron build configuration
https://www.electron.build/configuration/configuration
so you keep this inside your app's package.json.
I have the same simple built script for osx and windows
`"release": "electron-packager tmp --overwrite --out=release --asar --executableName=electron"`
it works fine on both os I mean I can run `npm run release` on both of the system and as an outcome I have bundled version

**Comment by devaffair** [comment]
So your solution indeed worked and I was able to publish!

However, a new issue occurred when trying to run the .exe file:
![image](https://user-images.githubusercontent.com/2995722/37485301-c886bd9a-2869-11e8-9dc6-f18c0bda0d88.png)

And I was wondering if you've encountered this issue before.

**Comment by obecny** [comment]
@devaffair all packages that need to be included in final package will be those that are installed as `dependencies`. I guess you included that as `devDependencies` ?

**Comment by devaffair** [comment]
It's located under "dependencies":
`"downstream-electron": "^0.2.0",`

After your reply, I added it to devDependencies, but it didn't change anything and the error is the same.

**Comment by khwaaj** [comment]
@devaffair, did this get resolved or are you still having problems?

**Comment by simongregory** [comment]
I've just updated to `v1.8.4-vmp1010` and ran the 64bit build to check our installer worked ok, and there weren't any problems again. I'd usually close the issue - but will leave it open for the second thread that's started up.

**Comment by khwaaj** [comment]
Since the original issue here has been resolved I'll close this. @devaffair, if you are still having problems with packaging downstream with your app, please open a new issue in the [downstream](https://github.com/castlabs/downstream_electron) repo.
